### ISSUE TITLE
[Fedora 32] create_engine:Exception

### ISSUE THREAD

**Issue by oatmealm** [mike-fabian/ibus-typing-booster#128]
I'm seeing this message login in to the default desktop (Wayland). Is Wayland supported?

`factory.py:84:do_create_engine:Exception: Cannot create engine typing-booster`

**Comment by mike-fabian** [comment]
Yes, Wayland is supported. I often test using Gnome-Wayland on Fedora 32.

There are some issues when using Wayland, colour in the preedit doesn’t work and there are some problems with surrounding text and with forward_key_event() (These details work better on Gnome-Xorg). But even on wayland, ibus-typing-booster works resonably well. 

Do you get this message always? It never works for you?

Is there anything interesting in the log file?:

```
~/.local/share/ibus-typing-booster/debug.log
```

**Comment by mike-fabian** [comment]
Does the problem always happen for you or only sometimes?

**Comment by oatmealm** [comment]
Hi @mike-fabian ... thanks for looking into it.

Yes, it's always the case. I was not able to yet see any layout with the booster enabled (i.e. XX + booster), though I see a generic typing booster layout, which works for English. I've enabled Spanish, Hebrew, Arabic dictionaries in the seetings for the booster.

Attaching the log.

<details><pre>
2020-07-20 10:40:01,156 main.py line 238 main INFO: ********** STARTING **********
2020-07-20 10:41:08,022 tabsqlitedb.py line 162 __init__ INFO: Compatible database /home/uri/.local/share/ibus-typing-booster/user.db found.
2020-07-20 10:41:08,022 tabsqlitedb.py line 171 __init__ INFO: Connect to the database /home/uri/.local/share/ibus-typing-booster/user.db.
2020-07-20 10:41:08,032 itb_util.py line 3295 _read_compose_file INFO: Reading compose file /usr/share/X11/locale/en_US.UTF-8/Compose
2020-07-20 10:41:08,228 itb_util.py line 3272 _read_compose_file INFO: Skipping reading of compose file "/home/uri/.config/ibus/Compose"
2020-07-20 10:41:08,228 itb_util.py line 3272 _read_compose_file INFO: Skipping reading of compose file "/home/uri/.XCompose"
2020-07-20 10:41:08,234 itb_util.py line 2897 get_hunspell_dictionary_wordlist INFO: /usr/share/myspell/he_IL.dic file found.
2020-07-20 10:41:08,235 itb_util.py line 2919 get_hunspell_dictionary_wordlist INFO: dictionary encoding=UTF-8 found in /usr/share/myspell/he_IL.aff
2020-07-20 10:41:08,306 itb_util.py line 2956 get_hunspell_dictionary_wordlist INFO: Successfully loaded /usr/share/myspell/he_IL.dic using UTF-8 encoding.
2020-07-20 10:41:09,118 itb_util.py line 2897 get_hunspell_dictionary_wordlist INFO: /usr/share/myspell/en_US.dic file found.
2020-07-20 10:41:09,118 itb_util.py line 2919 get_hunspell_dictionary_wordlist INFO: dictionary encoding=ISO8859-1 found in /usr/share/myspell/en_US.aff
2020-07-20 10:41:09,121 itb_util.py line 2956 get_hunspell_dictionary_wordlist INFO: Successfully loaded /usr/share/myspell/en_US.dic using ISO8859-1 encoding.
2020-07-20 10:41:09,301 hunspell_table.py line 552 __init__ INFO: ********** Initialized and ready for input: **********
2020-07-20 10:42:14,929 hunspell_table.py line 5083 on_gsettings_value_changed DEBUG: Settings changed: key=inputmethod value=he-kbd,NoIME

2020-07-20 10:42:33,241 hunspell_table.py line 5083 on_gsettings_value_changed DEBUG: Settings changed: key=inputmethod value=ar-kbd,he-kbd,NoIME

2020-07-20 10:46:22,301 main.py line 174 __bus_destroy_cb INFO: finalizing:)
2020-07-20 10:47:01,567 main.py line 238 main INFO: ********** STARTING **********
2020-07-20 10:47:01,690 tabsqlitedb.py line 162 __init__ INFO: Compatible database /home/uri/.local/share/ibus-typing-booster/user.db found.
2020-07-20 10:47:01,690 tabsqlitedb.py line 171 __init__ INFO: Connect to the database /home/uri/.local/share/ibus-typing-booster/user.db.
2020-07-20 10:47:01,718 itb_util.py line 3295 _read_compose_file INFO: Reading compose file /usr/share/X11/locale/en_US.UTF-8/Compose
2020-07-20 10:47:02,063 main.py line 174 __bus_destroy_cb INFO: finalizing:)
2020-07-20 10:47:03,036 main.py line 238 main INFO: ********** STARTING **********
2020-07-20 10:47:03,174 tabsqlitedb.py line 162 __init__ INFO: Compatible database /home/uri/.local/share/ibus-typing-booster/user.db found.
2020-07-20 10:47:03,175 tabsqlitedb.py line 171 __init__ INFO: Connect to the database /home/uri/.local/share/ibus-typing-booster/user.db.
2020-07-20 10:47:03,183 itb_util.py line 3295 _read_compose_file INFO: Reading compose file /usr/share/X11/locale/en_US.UTF-8/Compose
2020-07-20 10:47:03,516 itb_util.py line 3272 _read_compose_file INFO: Skipping reading of compose file "/home/uri/.config/ibus/Compose"
2020-07-20 10:47:03,516 itb_util.py line 3272 _read_compose_file INFO: Skipping reading of compose file "/home/uri/.XCompose"
2020-07-20 10:47:03,526 itb_util.py line 2897 get_hunspell_dictionary_wordlist INFO: /usr/share/myspell/he_IL.dic file found.
2020-07-20 10:47:03,527 itb_util.py line 2919 get_hunspell_dictionary_wordlist INFO: dictionary encoding=UTF-8 found in /usr/share/myspell/he_IL.aff
2020-07-20 10:47:03,661 itb_util.py line 2956 get_hunspell_dictionary_wordlist INFO: Successfully loaded /usr/share/myspell/he_IL.dic using UTF-8 encoding.
2020-07-20 10:47:04,711 itb_util.py line 2897 get_hunspell_dictionary_wordlist INFO: /usr/share/myspell/en_US.dic file found.
2020-07-20 10:47:04,711 itb_util.py line 2919 get_hunspell_dictionary_wordlist INFO: dictionary encoding=ISO8859-1 found in /usr/share/myspell/en_US.aff
2020-07-20 10:47:04,714 itb_util.py line 2956 get_hunspell_dictionary_wordlist INFO: Successfully loaded /usr/share/myspell/en_US.dic using ISO8859-1 encoding.
2020-07-20 10:47:04,905 hunspell_table.py line 552 __init__ INFO: ********** Initialized and ready for input: **********
2020-07-20 10:49:35,255 hunspell_table.py line 5083 on_gsettings_value_changed DEBUG: Settings changed: key=dictionary value=ar_SA,he_IL,en_US

2020-07-20 10:49:35,460 itb_util.py line 2897 get_hunspell_dictionary_wordlist INFO: /usr/share/myspell/ar_SA.dic file found.
2020-07-20 10:49:35,461 itb_util.py line 2919 get_hunspell_dictionary_wordlist INFO: dictionary encoding=UTF-8 found in /usr/share/myspell/ar_SA.aff
2020-07-20 10:49:35,539 itb_util.py line 2956 get_hunspell_dictionary_wordlist INFO: Successfully loaded /usr/share/myspell/ar_SA.dic using UTF-8 encoding.
2020-07-20 10:49:36,501 itb_util.py line 2897 get_hunspell_dictionary_wordlist INFO: /usr/share/myspell/he_IL.dic file found.
2020-07-20 10:49:36,501 itb_util.py line 2919 get_hunspell_dictionary_wordlist INFO: dictionary encoding=UTF-8 found in /usr/share/myspell/he_IL.aff
2020-07-20 10:49:36,566 itb_util.py line 2956 get_hunspell_dictionary_wordlist INFO: Successfully loaded /usr/share/myspell/he_IL.dic using UTF-8 encoding.
2020-07-20 10:49:37,357 itb_util.py line 2897 get_hunspell_dictionary_wordlist INFO: /usr/share/myspell/en_US.dic file found.
2020-07-20 10:49:37,357 itb_util.py line 2919 get_hunspell_dictionary_wordlist INFO: dictionary encoding=ISO8859-1 found in /usr/share/myspell/en_US.aff
2020-07-20 10:49:37,360 itb_util.py line 2956 get_hunspell_dictionary_wordlist INFO: Successfully loaded /usr/share/myspell/en_US.dic using ISO8859-1 encoding.
2020-07-20 10:56:39,062 main.py line 174 __bus_destroy_cb INFO: finalizing:)
</pre></details>

**Comment by mike-fabian** [comment]
Sometimes, it gets to 

```
********** Initialized and ready for input: **********
```
and then I think it works.

Like here in your log:

```
2020-07-20 10:41:09,301 hunspell_table.py line 552 __init__ INFO: ********** Initialized and ready for input: **********
2020-07-20 10:42:14,929 hunspell_table.py line 5083 on_gsettings_value_changed DEBUG: Settings changed: key=inputmethod value=he-kbd,NoIME

2020-07-20 10:42:33,241 hunspell_table.py line 5083 on_gsettings_value_changed DEBUG: Settings changed: key=inputmethod value=ar-kbd,he-kbd,NoIME

2020-07-20 10:46:22,301 main.py line 174 __bus_destroy_cb INFO: finalizing:)

2020-07-20 10:47:01,567 main.py line 238 main INFO: ********** STARTING **********
```

It was ready for input and then it apparently worked for a while, you changed some settings and maybe typed some stuff.
And then you probably logged out of Gnome and got the `__bus_destroy_cb INFO: finalizing:)`.
And at the next login to Gnome you got the `********** STARTING **********`.

But sometimes it never reaches the `********** Initialized and ready for input: **********`,
like here:

```
2020-07-20 10:47:01,567 main.py line 238 main INFO: ********** STARTING **********

2020-07-20 10:47:01,690 tabsqlitedb.py line 162 init INFO: Compatible database /home/uri/.local/share/ibus-typing-booster/user.db found.

2020-07-20 10:47:01,690 tabsqlitedb.py line 171 init INFO: Connect to the database /home/uri/.local/share/ibus-typing-booster/user.db.

2020-07-20 10:47:01,718 itb_util.py line 3295 _read_compose_file INFO: Reading compose file /usr/share/X11/locale/en_US.UTF-8/Compose

2020-07-20 10:47:02,063 main.py line 174 __bus_destroy_cb INFO: finalizing:)

2020-07-20 10:47:03,036 main.py line 238 main INFO: ********** STARTING **********
```

**Comment by mike-fabian** [comment]
Can you edit /usr/libexec/ibus-engine-typing-booster and change 

```
# Set this variable to something > 0 to get more debug output.
# (Debug output may show up in the log file and/or in the lookup table):
# export IBUS_TYPING_BOOSTER_DEBUG_LEVEL=1
```

to

```
export IBUS_TYPING_BOOSTER_DEBUG_LEVEL=255
```

Then there might be more output in the debug log.

And also set the debug level in the setup tool (where you added the dictionaries, in the “Options” tab to something > 0.

**Comment by mike-fabian** [comment]
![Screenshot](https://user-images.githubusercontent.com/2330175/88084868-4e1f8980-cb85-11ea-886a-656e14844ecf.png)

For me it seems to work fine. As you can see, I have configured he_IL, ar_SA, and en_US dictionaries and
the input methods ar-kbd, he-kbd, and NoIME.

NoIME means the keyboard input is used unchanged, i.e. if I type "a", I see "a" in the preedit.
ar-kbd simulates an Arabic keyboard on top of the US keyboard layout, i.e. if ar-kbd is at the top (highest priority), when I type "a", I see "ش"  in the preedit. And if he-kbd (which simulates a Hebrew keyboard on top of the US keyboard layout) is at the highest priority, typing the "a" key gives me a "ש" in the preedit.

All configured input methods are tried all the time, i.e. even if for example he-kbd is at the highest priority and you type a word which cannot be found in the he_IL dictionary but in the en_US dictionary, you might see Hebrew letters in the preedit but English candidates in the candidate list.

**Comment by mike-fabian** [comment]
The priority of the input methods can of course be changed in the setup tool with the buttons with the up and down arrows. And there are keyboard shortcuts (Control+Down and Control+Up) to change the priority of the input methods. For example, if NoIME is a the highest priority, you start typing and  see Latin Alphabet in the preedit and that is not what you wanted, you can type Control+Down to give the input method which was at the second highest priority the highest priority. If that was he-kbd, the preedit will switch to Hebrew letters.

**Comment by mike-fabian** [comment]
I just tried the above on Fedora 32 using Gnome Wayland and it worked fine for me.

**Comment by mike-fabian** [comment]
> I was not able to yet see any layout with the booster enabled (i.e. XX + booster), though I see a generic typing booster layout, which works for English.

Maybe there is just a misunderstanding how typing-booster works?

Typing-booster does not set a keyboard layout, it uses the keyboard layout which was used last.

For example, your keyboard layout could have been a US English layout before switching to typing-booster.

Then, in typing-booster, NoIME uses that layout just as it is, ar-kbd transliterates the Latin characters coming from the keyboard to Arabic characters, he-kbd transliterates the Latin characters to Hebrew characters (Select one of these input methods in the setup tool and click the "Input method help" button to see what a specific input method does.

**Comment by oatmealm** [comment]
@mike-fabian sorry but I'll get back to you with more info as soon as I can.

**Comment by mike-fabian** [comment]
ping 🏓

**Comment by oatmealm** [comment]
Hi , Sorry for the delay.
I think I'm not 100% sure how to set it up then. Do I need to remove my exsiting he/ar/en etc. layout and then add again looking for one that has "booster" (indicated in brackets)?

**Comment by mike-fabian** [comment]
No, you don’t need to remove any he/ar/en layouts. You just add ibus-typing-booster.

Here is the documentation on how to add it to your Gnome desktop:

http://mike-fabian.github.io/ibus-typing-booster/documentation.html#adding-gnome

Ibus-typing-booster is listed there as

    Other (Typing Booster)

See this screenshot:

http://mike-fabian.github.io/ibus-typing-booster/images/gnome-add-an-input-source-add-button.png

But I guess you already did that otherwise you would probably not have the debug.log you attached here.

Apparently you want to use it for Hebrew, Arabic, and English.

My previous comment has a screenshot which shows the setup tool of ibus-typing-booster and it is setup for Hebrew, Arabic and US English:

https://github.com/mike-fabian/ibus-typing-booster/issues/128#issuecomment-661990738

The text below the screenshot explains what this is for.

When you have it setup like this, you select the 'en' layout in your Gnome panel first, then select ibus-typing-booster in the Gnome panel (the rocket emoji: 🚀).

ibus-typing-booster uses the keyboard layout which already exists, i.e. it uses the 'en' layout now.

If you now type an 'a', this is interpreted in 3 ways:

1) NoIME (no input method) just take the a as it is

2) he-kbd (Hebrew keyboard simulation on top of a 'en' layout) transforms the 'a' into ש (HEBREW LETTER SHIN) because this is what a Hebrew keyboard has at the position where the 'en' keyboard has the 'a', see this picture:

https://en.wikipedia.org/wiki/Hebrew_keyboard#/media/File:Hebrew_keyboard_layout.png

3) ar-kbd (Arabic keyboard simulation on top of a 'en' layout) transforms the 'a' into ش (ARABIC LETTER SHEEN) because this is what an Arabic keyboard layout has at the position where the 'en' layout has the 'a'.

Ibus-typing-booster then tries to complete words using all three results (a ש ش).

'a' might give matches from the en_US dictionary.

'ש' might give matches from the he_IL dictionary.

'ش' might give matches from the ar_SA dictionary.

The matches are shown in the list of candidates as you can see in my screenshot above.

You may get matches from all the three dictionaries, so you may see matches in English, Hebrew and Arabic. 

Only one of the three typing results can be shown at the current writing position though.
Which one this is depends on the order of the entries “he-kbd, ar-kbd, and NoIME”.

You can change that order using the setup tool (not very convenient!) or quickly with the key binding Control + arrow up or Control + arrow down.

Now if you select 'he' in the Gnome panel and then select ibus-typing-booster, then
ibus-typing-booster will use the Hebrew layout. That means if you now type the key directly to the right of the CapsLock key, ibus-typing-booster will get a ש (HEBREW LETTER SHIN).

Again this letter is used in all three input methods:

1) NoIME takes it as it is, so you get ש

2) he-kbd doesn't understand ש and just passes it through so you get ש again

2) ar-kbd doesn't understand ש and just passes it through so you get ש again

So you get ש three times and see ש at your writing position.

All 3 dictionaries are tried to find a completion for ש. Nothing will be found in the en_US dictionary and the ar_SA dictionary, surely there will be no word starting with ש in these dictionaries. But there will be matches from the he_IL dictionary.

So now you get *only* Hebrew completion candidates. 

Now if you select 'ar' in the Gnome panel and then select ibus-typing-booster, then ibus-typing-booster will use the Arabic layout and similar as explained for Hebrew layout above, you will get *only* Arabic now.

If you prefer to switch languages by selecting 'en', 'ar', or 'he' in the Gnome panel and then returning to ibus-typing-booster, then you don’t need the he-kbd and ar-kbd input methods. As I explained above, they do nothing in that case and just waste a bit of time doing nothing. But if you do it this way, changing the language used by ibus-typing-booster requires two switches:

1) Switch from ibus-typing-booster to 'en', 'ar', or 'he' (using the Gnome panel)

2) Switch back to ibus-tying-booster

And, when you are in ibus-typing-booster, you have only one language.

Therefore, I think the more convenient way of using it is to have all three input methods 'NoIME', 'he-kbd', and 'ar-kbd' configured in ibus-typing-booster. Then select 'en' in the gnome panel and then select ibus-typing-booster.

Now you can stay in ibus-typing-booster and switch languages by using only one key press
(Control + arrow down or Control + arrow up).

And you may not even have to switch all the time because even when you are in Hebrew mode for example and try to input an English word, you will see Hebrew letters at the writing position but the English word you are trying to type will probably appear as a candidate and you can select it without switching away from Hebrew. I.e. you can switch faster and sometimes even avoid switching at all.

**Comment by oatmealm** [comment]
Works perfectly and looks VERY useful. Thanks a lot for the help, and sorry for the misunderstanding. I think I saw somewhere that after installing it, I should see in the list of available input source, something like "Hebrew (+booster) etc... I now understand I simply activate the booster from the panel input selection and it'll work on which ever layout is active, as long as it was configured in advance in the seeting for the booster input source itself.... Closing as it seems to work for me!
